Make joining decision pluggable, #144

patriknw opened the pull request:
* This centralized decision point also makes the actors easier to understand
* This probably fixes some issues with the old implementation also

Refs #144

ktoso commented on cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/JoinDecider.scala:
```diff
@@ -0,0 +1,208 @@
+    }
+
+  // FIXME Do we really need settings.contactPoint.noSeedsStableMargin? Isn't the stable-margin enough?
+  // It has been probing all known contact points for that duration
```
Yes I think you're right; those two can be collapsed into just using stable-margin for the decision point

ktoso commented on cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/dns/HeadlessServiceDnsBootstrap.scala:
```diff
@@ -279,16 +243,27 @@ final class HeadlessServiceDnsBootstrap(discovery: SimpleServiceDiscovery, setti
+      val info =
+        new SeedNodesInformation(currentTime, contacts.observedAt, contacts.observedContactPoints, seedObservations)
+
+      joinDecider.decide(info).recover { case _ => KeepProbing }.foreach(self ! _)
```
recover is good, but also log a warning then?

ktoso replied:
I guess no timeout is fine here hm; Technically we could end up invoking it again while it's still internally working on the previous decision future -- could lead to conflicting decisions? I guess we'd have acted upon the first thing already but maybe could show up as confusing logs if the decider decides to log things. 

Some more guard rails around this case would be nice... maybe if only logging that previous future decision still not completed etc

ktoso commented on cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/dns/HeadlessServiceDnsBootstrap.scala:
```diff
@@ -135,130 +103,126 @@ final class HeadlessServiceDnsBootstrap(discovery: SimpleServiceDiscovery, setti
   import HeadlessServiceDnsBootstrap._
   import context.dispatcher
 
+  // FIXME move/rename these classes, doesn't have to be DNS
```
yes good point Neither is it about Headless now as well, which is nice I think...? Headless is a specific k8s term, but all this is very generic now :)

ktoso commented:
Resolves https://github.com/akka/akka-management/issues/133 as well

ktoso commented on cluster-bootstrap/src/main/resources/reference.conf:
```diff
@@ -74,7 +74,7 @@ akka.management {
       # If some discovered seed node will keep failing to connect for specified period of time,
       # it will initiate rediscovery again instead of keep trying.
-      probing-failure-timeout = 5 seconds
+      probing-failure-timeout = 3 seconds
```
I think this needs some cleanup... The request coming back does not have a timeout it seems; yet this timeout is about "if we KEEP FAILING for this period, we'll resolve a new contact point and try again". I think we need to add a probe-timeout = 1.second perhaps and then should be good to have this at 3, so it's be 2/3 attempts after which we resolve again

patriknw commented:
Incorporated some of the feedback, in 56ec7ac
Will fix the other things also.

patriknw commented:
@ktoso Incorporated remaining feedback and updated documentation. This is ready for final review.

ktoso commented on cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/dns/HeadlessServiceDnsBootstrap.scala:
```diff
@@ -248,20 +251,34 @@ private[akka] final class HeadlessServiceDnsBootstrap(discovery: SimpleServiceDi
-      val currentTime = timeNow()
+  private def decide(): Unit = {
+    if (decisionInProgress)
+      log.debug("Previous decision still in progress")
```
Good though I was thinking warning; but debug is fine I guess

ktoso commented on cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/dns/HeadlessServiceDnsBootstrap.scala:
```diff
@@ -248,20 +251,34 @@ private[akka] final class HeadlessServiceDnsBootstrap(discovery: SimpleServiceDi
+          .recover {
+            case e =>
+              log.error(e, "Join decision failed: {}", e)
+              KeepProbing
```
good :)

ktoso commented on cluster-bootstrap/src/main/scala/akka/management/cluster/bootstrap/dns/HeadlessServiceDnsBootstrap.scala:
```diff
@@ -248,20 +251,34 @@ private[akka] final class HeadlessServiceDnsBootstrap(discovery: SimpleServiceDi
+          .decide(info)
+          .recover {
+            case e =>
+              log.error(e, "Join decision failed: {}", e)
```
e.getMessage) likely? Will change

patriknw commented:
@ktoso I fixed one thing that I had missed. `form-new-cluster` was lost in the refactoring, fixed in 3a6c9ee

I also moved the `LowestAddressJoinDecider` to a separate file as you suggested previously.

I'll do the other renaming of the classes in a separate PR, don't want to mix that with this diff.

ktoso commented:
Very nice, thanks!
